File: python/src/assetripper_import/platforms/ios_game_structure.py
"""Port of Source/AssetRipper.Import/Platforms/iOSGameStructure.cs"""
from __future__ import annotations

from ..structure.assembly.scripting_backend import ScriptingBackend
from .platform_game_structure import DATA_FOLDER_NAME, MANAGED_NAME, METADATA_NAME, RESOURCES_NAME, PlatformGameStructure

_IOS_STREAMING_NAME = "Raw"
_PAYLOAD_NAME = "Payload"
_APP_EXTENSION = ".app"


class iOSGameStructure(PlatformGameStructure):  # noqa: N801 -- matches Unity/upstream's own casing
    def __init__(self, root_path: str, file_system):
        super().__init__(file_system, root_path)

        found, data_path, app_path, name = _get_data_ios_directory(root_path, file_system)
        if not found:
            raise FileNotFoundError("Data directory wasn't found")

        self.name = name
        self.game_data_path = data_path
        self.streaming_assets_path = file_system.path.join(root_path, _IOS_STREAMING_NAME)
        self.resources_path = file_system.path.join(data_path, RESOURCES_NAME)
        self.managed_path = file_system.path.join(data_path, MANAGED_NAME)
        self.unity_player_path = None
        self.version = self._get_unity_version_from_data_directory(self.game_data_path)
        self.il2cpp_game_assembly_path = file_system.path.join(app_path, name)
        self.il2cpp_metadata_path = file_system.path.join(self.managed_path, METADATA_NAME, "global-metadata.dat")

        if self._has_il2cpp_files():
            self.backend = ScriptingBackend.IL2CPP
        elif self._has_mono_assemblies(self.managed_path):
            self.backend = ScriptingBackend.MONO
        else:
            self.backend = ScriptingBackend.UNKNOWN

        self.data_paths = [self.game_data_path]

    @staticmethod
    def exists(path: str, file_system) -> bool:
        if not file_system.directory.exists(path):
            return False
        found, _, _, _ = _get_data_ios_directory(path, file_system)
        return found


def _get_data_ios_directory(root_directory: str, file_system) -> tuple[bool, str | None, str | None, str | None]:
    payload_path = file_system.path.join(root_directory, _PAYLOAD_NAME)
    if not file_system.directory.exists(payload_path):
        return False, None, None, None

    for directory in file_system.directory.enumerate_directories(payload_path):
        name = file_system.path.get_file_name(directory)
        if name.endswith(_APP_EXTENSION):
            app_path = directory
            app_name = name[: -len(_APP_EXTENSION)]
            data_path = file_system.path.join(directory, DATA_FOLDER_NAME)
            if file_system.directory.exists(data_path):
                return True, data_path, app_path, app_name

    return False, None, None, None

File: Source/AssetRipper.Import/Platforms/iOSGameStructure.cs
using AssetRipper.Import.Structure.Assembly;
using AssetRipper.Import.Structure.Platforms;
using AssetRipper.IO.Files;

namespace AssetRipper.Import.Platforms;

internal sealed class iOSGameStructure : PlatformGameStructure
{
	public iOSGameStructure(string rootPath, FileSystem fileSystem) : base(rootPath, fileSystem)
	{
		if (!GetDataiOSDirectory(rootPath, FileSystem, out string? dataPath, out string? appPath, out string? name))
		{
			throw new DirectoryNotFoundException($"Data directory wasn't found");
		}

		Name = name;
		GameDataPath = dataPath;
		StreamingAssetsPath = FileSystem.Path.Join(rootPath, iOSStreamingName);
		ResourcesPath = FileSystem.Path.Join(dataPath, ResourcesName);
		ManagedPath = FileSystem.Path.Join(dataPath, ManagedName);
		UnityPlayerPath = null;
		Version = GetUnityVersionFromDataDirectory(GameDataPath);
		Il2CppGameAssemblyPath = GetIl2CppBinaryPath(appPath, name);
		Il2CppMetaDataPath = FileSystem.Path.Join(ManagedPath, MetadataName, DefaultGlobalMetadataName);

		if (HasIl2CppFiles())
		{
			Backend = ScriptingBackend.IL2Cpp;
		}
		else if (HasMonoAssemblies(ManagedPath))
		{
			Backend = ScriptingBackend.Mono;
		}
		else
		{
			Backend = ScriptingBackend.Unknown;
		}

		DataPaths = [GameDataPath];
	}

	public static bool Exists(string path, FileSystem fileSystem)
	{
		if (!fileSystem.Directory.Exists(path))
		{
			return false;
		}

		return GetDataiOSDirectory(path, fileSystem, out _, out _, out _);
	}

	private static bool GetDataiOSDirectory(string rootDirectory, FileSystem fileSystem, [NotNullWhen(true)] out string? dataPath, [NotNullWhen(true)] out string? appPath, [NotNullWhen(true)] out string? appName)
	{
		string payloadPath = fileSystem.Path.Join(rootDirectory, PayloadName);
		if (!fileSystem.Directory.Exists(payloadPath))
		{
			dataPath = null;
			appPath = null;
			appName = null;
			return false;
		}

		foreach (string directory in fileSystem.Directory.EnumerateDirectories(payloadPath))
		{
			string name = fileSystem.Path.GetFileName(directory);
			if (name.EndsWith(AppExtension, StringComparison.Ordinal))
			{
				appPath = directory;
				appName = name[..^AppExtension.Length];
				dataPath = fileSystem.Path.Join(directory, DataFolderName);
				if (fileSystem.Directory.Exists(dataPath))
				{
					return true;
				}
			}
		}

		dataPath = null;
		appPath = null;
		appName = null;
		return false;
	}

	/// <summary>
	/// AssetRipper: il2cpp code lives in UnityFramework.framework, not in the app executable.
	/// </summary>
	/// <remarks>
	/// Unity 2019.3 moved the player into an embedded framework, so the Mach-O named after the
	/// bundle is a launcher of a few tens of kilobytes with no managed code in it at all - on the
	/// test fixture, 70 KB against UnityFramework's 51 MB. Handing that one to LibCpp2IL fails with
	/// "No codegen modules found for mscorlib", which reads like a metadata problem and is not one:
	/// the metadata had already loaded, and the binary simply was not the game. Older builds do keep
	/// il2cpp in the app executable, so that stays the fallback.
	/// </remarks>
	private string GetIl2CppBinaryPath(string appPath, string name)
	{
		string frameworkPath = FileSystem.Path.Join(appPath, FrameworksName, UnityFrameworkName + FrameworkExtension, UnityFrameworkName);

		return FileSystem.File.Exists(frameworkPath)
			? frameworkPath
			: FileSystem.Path.Join(appPath, name);
	}

	private const string iOSStreamingName = "Raw";

	private const string PayloadName = "Payload";
	private const string AppExtension = ".app";
	private const string FrameworksName = "Frameworks";
	private const string UnityFrameworkName = "UnityFramework";
	private const string FrameworkExtension = ".framework";
}
